File: backend/core/src/main/java/site/timecapsulearchive/core/domain/capsule/generic_capsule/data/dto/CapsuleSummaryDto.java
package site.timecapsulearchive.core.domain.capsule.generic_capsule.data.dto;

import java.time.ZonedDateTime;
import java.util.function.Function;
import org.locationtech.jts.geom.Point;
import site.timecapsulearchive.core.domain.capsule.generic_capsule.data.response.CapsuleSummaryResponse;

public record CapsuleSummaryDto(
    String nickname,
    String profileUrl,
    String skinUrl,
    String title,
    ZonedDateTime dueDate,
    Point point,
    String address,
    String roadName,
    Boolean isOpened,
    ZonedDateTime createdAt
) {

    public CapsuleSummaryResponse toResponse(
        final Function<String, String> preSignUrlFunction,
        final Function<Point, Point> changePointFunction
    ) {
        final Point changePoint = changePointFunction.apply(point);

        return CapsuleSummaryResponse.builder()
            .nickname(nickname)
            .profileUrl(profileUrl)
            .skinUrl(preSignUrlFunction.apply(skinUrl))
            .title(title)
            .dueDate(dueDate)
            .latitude(changePoint.getX())
            .longitude(changePoint.getY())
            .address(address)
            .roadName(roadName)
            .isOpened(isOpened)
            .createdAt(createdAt)
            .build();
    }

}

File: frontend/ARchive/app/src/main/java/com/droidblossom/archive/data/dto/common/CapsuleSummaryDto.kt
package com.droidblossom.archive.data.dto.common

import com.droidblossom.archive.domain.model.capsule.CapsuleAnchor
import com.droidblossom.archive.domain.model.capsule.CapsuleMarker
import com.droidblossom.archive.presentation.ui.home.HomeFragment

data class CapsuleSummaryDto(
    val id: Long,
    val longitude: Double,
    val latitude: Double,
    val nickname: String,
    val capsuleSkinUrl: String,
    val title: String,
    val dueDate: String,
    val capsuleType: String,
){
    fun toAnchorModel() = CapsuleAnchor(
        id = this.id,
        longitude = this.longitude,
        latitude = this.latitude,
        capsuleType = HomeFragment.CapsuleType.valueOf(capsuleType),
        skinUrl = this.capsuleSkinUrl
    )

    fun toMarkerModel() = CapsuleMarker(
        id = this.id,
        longitude = this.longitude,
        latitude = this.latitude,
        capsuleType = HomeFragment.CapsuleType.valueOf(capsuleType),
    )
}
